<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Vivarium E. coli Processes &mdash; Vivarium E. coli 0.0.1 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Vivarium E. coli Composites" href="Vivarium_ecoli_composites.html" />
    <link rel="prev" title="Tutorials" href="../tutorials.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Vivarium E. coli
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../reference/processes.html">Processes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/composites.html">Composites</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/experiments.html">Experiments</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Vivarium E. coli Processes</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Load-the-required-components">Load the required components</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#import-modules">import modules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Load-sim_data">Load sim_data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Get-initial-state-snapshot">Get initial state snapshot</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Helper-functions-for-executing-the-notebook">Helper functions for executing the notebook</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Complexation">Complexation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Transcript-Initiation">Transcript Initiation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Transcript-Elongation">Transcript Elongation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Transcription-Factor-Binding">Transcription Factor Binding</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Chromosome-Replication">Chromosome Replication</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Chromosome-Structure">Chromosome Structure</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Polypeptide-Initiation">Polypeptide Initiation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Polypeptide-Elongation">Polypeptide Elongation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Protein-Degradation">Protein Degradation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#RNA-Degradation">RNA Degradation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Two-Component-System">Two Component System</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Equilibrium">Equilibrium</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Metabolism">Metabolism</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Vivarium_ecoli_composites.html">Vivarium E. coli Composites</a></li>
<li class="toctree-l2"><a class="reference internal" href="allocation_processes.html">Allocation using PartitionedProcess</a></li>
<li class="toctree-l2"><a class="reference internal" href="ecoli_simulations.html"><em>E. coli</em> Simulations</a></li>
<li class="toctree-l2"><a class="reference internal" href="colony_simulations.html">Colony Simulations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../docs.html">Working with Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Material</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Vivarium E. coli</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../tutorials.html">Tutorials</a> &raquo;</li>
      <li>Vivarium E. coli Processes</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/notebooks/Vivarium_ecoli_processes.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="Vivarium-E.-coli-Processes">
<h1>Vivarium E. coli Processes<a class="headerlink" href="#Vivarium-E.-coli-Processes" title="Permalink to this heading"></a></h1>
<p>This notebook demonstrates how the processes work individually. For each process, we will:</p>
<ol class="arabic simple">
<li><p>Load in the process and parameters</p></li>
<li><p>Plot a <strong>toplogy</strong> diagram</p>
<ul class="simple">
<li><p>The topology is a network that demonstrates how a process connects to its stores (which hold state variables).</p></li>
</ul>
</li>
<li><p>Display the <strong>ports schema</strong></p>
<ul class="simple">
<li><p>The port schema defines a systems ports (top-level keys), and the expected behavior of molecules under that port (its <em>schema</em>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">*</span></code> is a wild card, specifies the schema of everything that can go into the port</p></li>
</ul>
</li>
<li><p>Simulate the process</p></li>
<li><p>Demonstrate distinct features of that process</p></li>
</ol>
<p><strong>Miscellaneous notes/ideas:</strong> - consider adding interactive widgets (plotly) - users can click boxes to choose which molecules to plot</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make sure notebook runs out of vivarium-ecoli directory</span>
<span class="kn">import</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">os</span>

<span class="c1"># get the path to the notebook, and change working directory</span>
<span class="n">notebook_path</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;notebooks&#39;</span><span class="p">)]</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">notebook_path</span><span class="p">)</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>
</pre></div>
</div>
</div>
<section id="Load-the-required-components">
<h2>Load the required components<a class="headerlink" href="#Load-the-required-components" title="Permalink to this heading"></a></h2>
<section id="import-modules">
<h3>import modules<a class="headerlink" href="#import-modules" title="Permalink to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">vivarium.core.process</span> <span class="kn">import</span> <span class="n">Process</span>
<span class="kn">from</span> <span class="nn">vivarium.core.store</span> <span class="kn">import</span> <span class="n">Store</span>
<span class="kn">from</span> <span class="nn">vivarium.core.engine</span> <span class="kn">import</span> <span class="n">Engine</span><span class="p">,</span> <span class="n">pp</span>
<span class="kn">from</span> <span class="nn">vivarium.core.composition</span> <span class="kn">import</span> <span class="n">simulate_process</span><span class="p">,</span> <span class="n">simulate_composite</span>
<span class="kn">from</span> <span class="nn">vivarium.plots.topology</span> <span class="kn">import</span> <span class="n">plot_topology</span>
<span class="kn">from</span> <span class="nn">vivarium.plots.simulation_output</span> <span class="kn">import</span> <span class="n">plot_variables</span>
<span class="kn">from</span> <span class="nn">ecoli.processes.registries</span> <span class="kn">import</span> <span class="n">topology_registry</span>
<span class="kn">import</span> <span class="nn">ecoli</span>
<span class="kn">import</span> <span class="nn">copy</span>
</pre></div>
</div>
</div>
</section>
<section id="Load-sim_data">
<h3>Load sim_data<a class="headerlink" href="#Load-sim_data" title="Permalink to this heading"></a></h3>
<p>This unpickles and <code class="docutils literal notranslate"><span class="pre">sim_data</span></code> object, which holds model parameters from wcEcoli. <code class="docutils literal notranslate"><span class="pre">LoadSimData</span></code> includes functions to retrieve individual processes’ parameters, which can be modified and passed into their respective process models.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.library.sim_data</span> <span class="kn">import</span> <span class="n">LoadSimData</span>

<span class="n">SIM_DATA_PATH</span> <span class="o">=</span> <span class="s1">&#39;reconstruction/sim_data/kb/simData.cPickle&#39;</span>

<span class="n">load_sim_data</span> <span class="o">=</span> <span class="n">LoadSimData</span><span class="p">(</span>
            <span class="n">sim_data_path</span><span class="o">=</span><span class="n">SIM_DATA_PATH</span><span class="p">,</span>
            <span class="n">seed</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Get-initial-state-snapshot">
<h3>Get initial state snapshot<a class="headerlink" href="#Get-initial-state-snapshot" title="Permalink to this heading"></a></h3>
<p><code class="docutils literal notranslate"><span class="pre">initial_state</span></code> is a dict with the initial state of the system – a snapshot saved from wcEcoli.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.composites.ecoli_master</span> <span class="kn">import</span> <span class="n">get_state_from_file</span>

<span class="n">INITIAL_STATE_PATH</span> <span class="o">=</span> <span class="s1">&#39;data/wcecoli_t1000.json&#39;</span>

<span class="n">initial_state</span> <span class="o">=</span> <span class="n">get_state_from_file</span><span class="p">(</span><span class="n">path</span><span class="o">=</span><span class="n">INITIAL_STATE_PATH</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Helper-functions-for-executing-the-notebook">
<h3>Helper functions for executing the notebook<a class="headerlink" href="#Helper-functions-for-executing-the-notebook" title="Permalink to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">schema_keys</span> <span class="o">=</span> <span class="n">Store</span><span class="o">.</span><span class="n">schema_keys</span>

<span class="k">def</span> <span class="nf">make_port_printout</span><span class="p">(</span><span class="n">ports_schema</span><span class="p">,</span> <span class="n">depth</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">schema_show</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">filler_size</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
    <span class="n">print_dict</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="n">filler</span> <span class="o">=</span> <span class="n">filler_size</span> <span class="o">*</span> <span class="s1">&#39; &#39;</span>
    <span class="k">for</span> <span class="n">port</span><span class="p">,</span> <span class="n">schema</span> <span class="ow">in</span> <span class="n">ports_schema</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">schema</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>
            <span class="n">schemavars</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">schema</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">var</span> <span class="ow">in</span> <span class="n">schemavars</span> <span class="k">for</span> <span class="n">var</span> <span class="ow">in</span> <span class="n">schema_keys</span><span class="p">):</span>
                <span class="n">print_schema</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">schema</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="n">print_schema</span> <span class="o">+=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="p">(</span><span class="n">depth</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">filler</span><span class="si">}</span><span class="s1"> </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s1">: </span><span class="si">{</span><span class="n">v</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span>
                <span class="n">print_dict</span> <span class="o">+=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">depth</span> <span class="o">*</span> <span class="n">filler</span><span class="si">}{</span><span class="n">port</span><span class="si">}</span><span class="s1">:</span><span class="se">\n</span><span class="si">{</span><span class="n">print_schema</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">schema_items</span> <span class="o">=</span> <span class="n">schema</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                <span class="n">first_schema</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">schema_items</span><span class="p">)[:</span><span class="n">schema_show</span><span class="p">])</span>
                <span class="n">next_print</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">first_schema</span><span class="p">,</span> <span class="n">depth</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">print_dict</span> <span class="o">+=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">port</span><span class="si">}</span><span class="s1">:</span><span class="se">\n</span><span class="si">{</span><span class="n">next_print</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">schema</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">schema_show</span><span class="p">:</span>
                    <span class="n">print_dict</span> <span class="o">+=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="p">(</span><span class="n">depth</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">filler</span><span class="si">}</span><span class="s1">&#39;</span>
                    <span class="n">print_dict</span> <span class="o">+=</span> <span class="sa">f</span><span class="s1">&#39;... skipping </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">schema</span><span class="p">)</span><span class="o">-</span><span class="n">schema_show</span><span class="si">}</span><span class="s1"> schema entries ...&#39;</span>
                    <span class="n">print_dict</span> <span class="o">+=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="se">\n\n</span><span class="s1">&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">print_dict</span> <span class="o">+=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">filler</span><span class="si">}{</span><span class="n">schema</span><span class="si">}</span><span class="se">\n</span><span class="s1">&#39;</span>
    <span class="k">return</span> <span class="n">print_dict</span>

<span class="k">def</span> <span class="nf">find_increasing</span><span class="p">(</span><span class="n">d</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">value</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">return</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">value</span><span class="p">}</span>

<span class="k">def</span> <span class="nf">find_decreasing</span><span class="p">(</span><span class="n">d</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">value</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">return</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">value</span><span class="p">}</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Complexation">
<h2>Complexation<a class="headerlink" href="#Complexation" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.complexation.html">Complexation Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.complexation</span> <span class="kn">import</span> <span class="n">Complexation</span>

<span class="c1"># load in parameters</span>
<span class="n">cplx_config</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_complexation_config</span><span class="p">()</span>
<span class="n">cplx_config</span><span class="p">[</span><span class="s1">&#39;time_step&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.01</span>

<span class="c1"># initialize process</span>
<span class="n">complexation</span> <span class="o">=</span> <span class="n">Complexation</span><span class="p">(</span><span class="n">cplx_config</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">cplx_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;buffer&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-complexation&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">complexation&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">10000</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span>
<span class="p">}</span>

<span class="n">cplx_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">complexation</span><span class="p">,</span> <span class="n">cplx_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_13_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_13_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">cplx_ports</span> <span class="o">=</span> <span class="n">complexation</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">cplx_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">cplx_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cplx_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
molecules:
     1-PFK[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     1-PFK-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2OXOGLUTARATEDEH-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     E1O[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     E2O[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 2569 schema entries ...

listeners:
     complexation_events:
           _default: []
           _updater: set
           _emit: True



</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># tweak initial state</span>
<span class="n">cplx_initial_state</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">initial_state</span><span class="p">)</span>
<span class="n">cplx_initial_state</span><span class="p">[</span><span class="s1">&#39;bulk&#39;</span><span class="p">][</span><span class="s1">&#39;6PGLUCONDEHYDROG-MONOMER[c]&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1000</span>

<span class="c1"># run simulation and retrieve final data</span>
<span class="n">cplx_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">cplx_initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">complexation</span><span class="o">.</span><span class="n">topology</span><span class="p">}</span>

<span class="n">cplx_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">complexation</span><span class="p">,</span> <span class="n">cplx_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: a7b0130c-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:04
Completed in 0.178275 seconds
</pre></div></div>
</div>
<p>For complexation, let’s look at the 6PGLUCONDEHYDROG-MONOMER[c] monomer as it transitions to the 6PGLUCONDEHYDROG-CPLX[c] complex:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot output</span>
<span class="n">cplx_fig</span> <span class="o">=</span> <span class="n">plot_variables</span><span class="p">(</span>
    <span class="n">cplx_data</span><span class="p">,</span>
    <span class="n">variables</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;6PGLUCONDEHYDROG-MONOMER[c]&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;6PGLUCONDEHYDROG-CPLX[c]&#39;</span><span class="p">),</span>
    <span class="p">],</span>
    <span class="n">column_width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">row_height</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">row_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_17_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_17_0.png" />
</div>
</div>
<p>Here we see 6PGLUCONDEHYDROG-MONOMER[c] getting complexed. This a relatively fast process and consumes all the monomers in a single time step.</p>
</section>
<section id="Transcript-Initiation">
<h2>Transcript Initiation<a class="headerlink" href="#Transcript-Initiation" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.transcript_initiation.html">Transcript Initiation Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.transcript_initiation</span> <span class="kn">import</span> <span class="n">TranscriptInitiation</span>

<span class="c1"># load in parameters</span>
<span class="n">ti_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_transcript_initiation_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">transcript_initiation</span> <span class="o">=</span> <span class="n">TranscriptInitiation</span><span class="p">(</span><span class="n">ti_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">ti_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-transcript-initiation&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">transcript</span><span class="se">\n</span><span class="s1">initiation&#39;</span><span class="p">,</span>
        <span class="s1">&#39;full_chromosomes&#39;</span><span class="p">:</span> <span class="s1">&#39;full</span><span class="se">\n</span><span class="s1">chromosomes&#39;</span><span class="p">,</span>
        <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">rna_synth_prob&#39;</span><span class="p">:</span> <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">rna_synth_</span><span class="se">\n</span><span class="s1">prob&#39;</span><span class="p">,</span>
        <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">ribosome_data&#39;</span><span class="p">:</span> <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">ribosome_</span><span class="se">\n</span><span class="s1">data&#39;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">17000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">3.3</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">18</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-transcript-initiation&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">ti_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">transcript_initiation</span><span class="p">,</span> <span class="n">ti_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_21_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_21_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">ti_ports</span> <span class="o">=</span> <span class="n">transcript_initiation</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">ti_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">ti_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ti_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
environment:
     media_id:
           _default:
           _updater: set


molecules:
     APORNAP-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     GUANOSINE-5DP-3DP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


full_chromosomes:
*:
          unique_index:
                _default: 0



promoters:
*:
          TU_index:
                _default: 0

          coordinates:
                _default: 0

          domain_index:
                _default: 0

          bound_TF:
                _default: 0



RNAs:
*:
          unique_index:
                _default: 0
                _updater: set

          TU_index:
                _default: 0
                _updater: set

          transcript_length:
                _default: 0
                _updater: set
                _emit: True

          is_mRNA:
                _default: 0
                _updater: set

          is_full_transcript:
                _default: 0
                _updater: set


          ... skipping 2 schema entries ...


active_RNAPs:
*:
          unique_index:
                _default: 0
                _updater: set

          domain_index:
                _default: 0
                _updater: set

          coordinates:
                _default: 0
                _updater: set
                _emit: True

          direction:
                _default: 0
                _updater: set



listeners:
mass:
          cell_mass:
                _default: 0.0

          dry_mass:
                _default: 0.0


rna_synth_prob:
          rna_synth_prob:
                _default: 0.0
                _updater: set
                _emit: True


ribosome_data:
          rrn16S_produced:
                _default: 0
                _updater: set
                _emit: True

          rrn23S_produced:
                _default: 0
                _updater: set
                _emit: True

          rrn5S_produced:
                _default: 0
                _updater: set
                _emit: True

          rrn16S_init_prob:
                _default: 0
                _updater: set
                _emit: True

          rrn23S_init_prob:
                _default: 0
                _updater: set
                _emit: True


          ... skipping 2 schema entries ...

rnap_data:
          didInitialize:
                _default: 0
                _updater: set
                _emit: True

          rnaInitEvent:
                _default: 0
                _updater: set
                _emit: True




</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run simulation and retrieve final data</span>
<span class="n">ti_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">transcript_initiation</span><span class="o">.</span><span class="n">topology</span><span class="p">}</span>

<span class="n">ti_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">transcript_initiation</span><span class="p">,</span> <span class="n">ti_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: a81ba680-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:05
Completed in 1.46 seconds
</pre></div></div>
</div>
<p>For Transcript Initiation, we can see from the cell above that each active RNA polymerase molecule is represented by an ID number. Let’s analyze how one of these active RNA polymerase molecules functions within this process:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot output</span>
<span class="n">RNAP_ID</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">ti_data</span><span class="p">[</span><span class="s1">&#39;unique&#39;</span><span class="p">][</span><span class="s1">&#39;active_RNAP&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span>

<span class="n">ti_fig</span> <span class="o">=</span> <span class="n">plot_variables</span><span class="p">(</span>
    <span class="n">ti_data</span><span class="p">,</span>
    <span class="n">variables</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s1">&#39;unique&#39;</span><span class="p">,</span> <span class="s1">&#39;active_RNAP&#39;</span><span class="p">,</span> <span class="n">RNAP_ID</span><span class="p">,</span> <span class="s1">&#39;coordinates&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;APORNAP-CPLX[c]&#39;</span><span class="p">)</span>
        <span class="p">],</span>
    <span class="n">column_width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">row_height</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">row_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_25_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_25_0.png" />
</div>
</div>
<p>Here we can see that the coordinates for one RNA polymerase molecule is initialized at time=0 and remains the same throughout the simulation as elongation is not a function of this process. Additionally, we can see that the RNA polymerase molecules (given by APORNAP-CPLX[c]) are getting deleted from the bulk molecules count as they bind to the DNA sequence and are initialized for transcription.</p>
</section>
<section id="Transcript-Elongation">
<h2>Transcript Elongation<a class="headerlink" href="#Transcript-Elongation" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.transcript_elongation.html">Transcript Elongation Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.transcript_elongation</span> <span class="kn">import</span> <span class="n">TranscriptElongation</span>

<span class="c1"># load in parameters</span>
<span class="n">te_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_transcript_elongation_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">transcript_elongation</span> <span class="o">=</span> <span class="n">TranscriptElongation</span><span class="p">(</span><span class="n">te_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">te_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-transcript-elongation&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">transcript</span><span class="se">\n</span><span class="s1">elongation&#39;</span><span class="p">,</span>
        <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">transcript_elongation_listener&#39;</span><span class="p">:</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">listeners</span><span class="se">\n</span><span class="s1">transcript_</span><span class="se">\n</span><span class="s1">elongation_</span><span class="se">\n</span><span class="s1">listener&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">17000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">3.3</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">18</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-transcript-elongation&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">te_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">transcript_elongation</span><span class="p">,</span> <span class="n">te_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_29_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_29_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">te_ports</span> <span class="o">=</span> <span class="n">transcript_elongation</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">te_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">te_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">te_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
environment:
     media_id:
           _default:


RNAs:
_divider:
     &lt;function divide_RNAs_by_domain at 0x12dcc89d0&gt;
topology:
     (&#39;..&#39;, &#39;active_RNAP&#39;)


*:
          unique_index:
                _default: 0
                _updater: set

          TU_index:
                _default: 0
                _updater: set

          transcript_length:
                _default: 0
                _updater: set
                _emit: True

          is_mRNA:
                _default: False
                _updater: set

          is_full_transcript:
                _default: False
                _updater: set


          ... skipping 3 schema entries ...


active_RNAPs:
_divider:
     &lt;function divide_active_RNAPs_by_domain at 0x12dcc88b0&gt;

*:
          unique_index:
                _default: 0
                _updater: set

          domain_index:
                _default: 0
                _updater: set

          coordinates:
                _default: 0
                _updater: set
                _emit: True

          direction:
                _default: 0
                _updater: set



bulk_RNAs:
     EG10001_RNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10002_RNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10003_RNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10004_RNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10006_RNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 4682 schema entries ...

ntps:
     ATP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     GTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     UTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


molecules:
     PPI[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     APORNAP-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


listeners:
mass:
          cell_mass:
                _default: 0.0


transcript_elongation_listener:
          countNTPsUsed:
                _default: 0
                _updater: set
                _emit: True

          countRnaSynthesized:
                _default: [0. 0. 0. ... 0. 0. 0.]
                _updater: set
                _emit: True

          attenuation_probability:
                _updater: set
                _emit: True

          counts_attenuated:
                _updater: set
                _emit: True


growth_limits:
          ntpUsed:
                _default: [0. 0. 0. 0.]
                _updater: set
                _emit: True


rnap_data:
          actualElongations:
                _default: 0
                _updater: set
                _emit: True

          didTerminate:
                _default: 0
                _updater: set
                _emit: True

          terminationLoss:
                _default: 0
                _updater: set
                _emit: True

          didStall:
                _updater: set
                _emit: True




</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run simulation and retrieve final data</span>
<span class="n">te_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">transcript_elongation</span><span class="o">.</span><span class="n">topology</span><span class="p">}</span>

<span class="n">te_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">transcript_elongation</span><span class="p">,</span> <span class="n">te_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: aa11a2fa-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:09
Completed in 2.21 seconds
</pre></div></div>
</div>
<p>For Transcript Elongation, we can see from the cell above that each active RNA Polymerase molecule is represented by an ID number. Let’s analyze how a few of these active RNA Polymerase molecules function within this process:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot output</span>
<span class="n">te_fig</span> <span class="o">=</span> <span class="n">plot_variables</span><span class="p">(</span>
    <span class="n">te_data</span><span class="p">,</span>
    <span class="n">variables</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s1">&#39;unique&#39;</span><span class="p">,</span> <span class="s1">&#39;active_RNAP&#39;</span><span class="p">,</span> <span class="s1">&#39;1266660&#39;</span><span class="p">,</span> <span class="s1">&#39;coordinates&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;unique&#39;</span><span class="p">,</span> <span class="s1">&#39;active_RNAP&#39;</span><span class="p">,</span> <span class="s1">&#39;1293463&#39;</span><span class="p">,</span> <span class="s1">&#39;coordinates&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;unique&#39;</span><span class="p">,</span> <span class="s1">&#39;active_RNAP&#39;</span><span class="p">,</span> <span class="s1">&#39;1293466&#39;</span><span class="p">,</span> <span class="s1">&#39;coordinates&#39;</span><span class="p">)</span>
        <span class="p">],</span>
    <span class="n">column_width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">row_height</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">row_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_33_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_33_0.png" />
</div>
</div>
<p>Here we can see that the coordinates for these RNA Polymerase molecules are initialized at time=0 but change throughout the simulation. Some polymerase coordinates incease, indicating elongation in one direction, and others decrease, indicating elongation in the opposite direction along the DNA sequence.</p>
<p>Funadamentally, these changes represent the process of polymerization: as the RNA polymerase molecules travel across the DNA, RNA molecucles are assembled.</p>
</section>
<section id="Transcription-Factor-Binding">
<h2>Transcription Factor Binding<a class="headerlink" href="#Transcription-Factor-Binding" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.tf_binding.html">TF Binding Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.tf_binding</span> <span class="kn">import</span> <span class="n">TfBinding</span>

<span class="c1"># load in parameters</span>
<span class="n">tfb_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_tf_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">tf_binding</span> <span class="o">=</span> <span class="n">TfBinding</span><span class="p">(</span><span class="n">tfb_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">tfb_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-tf-binding&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">tf binding&#39;</span><span class="p">,</span>
        <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">rna_synth_prob&#39;</span><span class="p">:</span> <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">rna_synth_</span><span class="se">\n</span><span class="s1">prob&#39;</span><span class="p">,</span>
        <span class="s1">&#39;active_tfs_total&#39;</span><span class="p">:</span> <span class="s1">&#39;active_tfs_</span><span class="se">\n</span><span class="s1">total&#39;</span><span class="p">,</span>
        <span class="s1">&#39;inactive_tfs_total&#39;</span><span class="p">:</span> <span class="s1">&#39;inactive_tfs_</span><span class="se">\n</span><span class="s1">total&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">16000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">3.5</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">18</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-tf-binding&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">tfb_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">tf_binding</span><span class="p">,</span> <span class="n">tfb_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_37_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_37_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">tfb_ports</span> <span class="o">=</span> <span class="n">tf_binding</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">tfb_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">tfb_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tfb_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
promoters:
*:
          TU_index:
                _default: 0
                _updater: set
                _emit: True

          bound_TF:
                _default: 0
                _updater: set
                _emit: True

          submass:
                _default: [0. 0. 0. 0. 0. 0. 0. 0. 0.]
                _emit: True



active_tfs:
     CPLX-125[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX-172[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-226[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-228[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-7669[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 19 schema entries ...

active_tfs_total:
     CPLX-125[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     CPLX-172[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     CPLX0-226[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     CPLX0-228[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     CPLX0-7669[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True


     ... skipping 19 schema entries ...

inactive_tfs_total:
     PC00007[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     PC00003[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     PC00004[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     PC00005[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     CPLX0-7670[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True


     ... skipping 12 schema entries ...

listeners:
rna_synth_prob:
          pPromoterBound:
                _default: 0
                _updater: set
                _emit: True

          nPromoterBound:
                _default: 0
                _updater: set
                _emit: True

          nActualBound:
                _default: 0
                _updater: set
                _emit: True

          n_available_promoters:
                _default: 0
                _updater: set
                _emit: True

          n_bound_TF_per_TU:
                _default: 0
                _updater: set
                _emit: True


          ... skipping 1 schema entries ...



</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run simulation and retrieve final data</span>
<span class="n">tfb_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">tf_binding</span><span class="o">.</span><span class="n">topology</span><span class="p">}</span>

<span class="n">tfb_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">tf_binding</span><span class="p">,</span> <span class="n">tfb_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: acb463e4-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:12
Completed in 2.07 seconds
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot output</span>
<span class="n">tfb_fig_active</span> <span class="o">=</span> <span class="n">plot_variables</span><span class="p">(</span>
    <span class="n">tfb_data</span><span class="p">,</span>
    <span class="n">variables</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;CPLX-125[c]&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;CPLX0-226[c]&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;MONOMER0-162[c]&#39;</span><span class="p">)</span>
        <span class="p">],</span>
    <span class="n">column_width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">row_height</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">row_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_40_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_40_0.png" />
</div>
</div>
</section>
<section id="Chromosome-Replication">
<h2>Chromosome Replication<a class="headerlink" href="#Chromosome-Replication" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.chromosome_replication.html">Chromosome Replication Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.chromosome_replication</span> <span class="kn">import</span> <span class="n">ChromosomeReplication</span>

<span class="c1"># load in parameters</span>
<span class="n">cr_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_chromosome_replication_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">chromosome_replication</span> <span class="o">=</span> <span class="n">ChromosomeReplication</span><span class="p">(</span><span class="n">cr_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">cr_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-chromosome-replication&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">chromosome</span><span class="se">\n</span><span class="s1">replication&#39;</span><span class="p">,</span>
        <span class="s1">&#39;replisome_trimers&#39;</span><span class="p">:</span> <span class="s1">&#39;replisome</span><span class="se">\n</span><span class="s1">trimers&#39;</span><span class="p">,</span>
        <span class="s1">&#39;replisome_monomers&#39;</span><span class="p">:</span> <span class="s1">&#39;replisome</span><span class="se">\n</span><span class="s1">monomers&#39;</span><span class="p">,</span>
        <span class="s1">&#39;active_replisomes&#39;</span><span class="p">:</span> <span class="s1">&#39;active</span><span class="se">\n</span><span class="s1">replisomes&#39;</span><span class="p">,</span>
        <span class="s1">&#39;full_chromosomes&#39;</span><span class="p">:</span> <span class="s1">&#39;full</span><span class="se">\n</span><span class="s1">chromosomes&#39;</span><span class="p">,</span>
        <span class="s1">&#39;chromosome_domains&#39;</span><span class="p">:</span> <span class="s1">&#39;chromosome</span><span class="se">\n</span><span class="s1">domains&#39;</span><span class="p">,</span>
        <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">replication_data&#39;</span><span class="p">:</span> <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">replication_</span><span class="se">\n</span><span class="s1">data&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">17000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">3.5</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">18</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-chromosome-replication&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">cr_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">chromosome_replication</span><span class="p">,</span> <span class="n">cr_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_43_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_43_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">cr_ports</span> <span class="o">=</span> <span class="n">chromosome_replication</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">cr_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">cr_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cr_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
replisome_trimers:
     CPLX0-2361[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-3761[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


replisome_monomers:
     CPLX0-3621[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10239-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG11500-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG11412-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


dntps:
     DATP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     DCTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     DGTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     TTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


ppi:
     PPI[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


listeners:
mass:
          cell_mass:
                _default: 0.0
                _emit: True


replication_data:
          criticalInitiationMass:
                _default: 0.0

          criticalMassPerOriC:
                _default: 0.0



environment:
     media_id:
           _default:
           _updater: set


active_replisomes:
*:
          domain_index:
                _default: 0
                _updater: set
                _emit: True

          right_replichore:
                _default: 0
                _updater: set
                _emit: True

          coordinates:
                _default: 0
                _updater: set
                _emit: True

          submass:
                _default: [0. 0. 0. 0. 0. 0. 0. 0. 0.]
                _emit: True



oriCs:
*:
          domain_index:
                _default: 0
                _updater: set
                _emit: True



chromosome_domains:
*:
          domain_index:
                _default: 0
                _updater: set
                _emit: True

          child_domains:
                _default: 0
                _updater: set
                _emit: True



full_chromosomes:
*:
          domain_index:
                _default: 0
                _updater: set
                _emit: True

          division_time:
                _default: 0
                _updater: set
                _emit: True

          has_triggered_division:
                _default: False
                _updater: set




</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># tweak initial state to trigger replication</span>
<span class="n">cr_initial_state</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">initial_state</span><span class="p">)</span>
<span class="n">cr_initial_state</span><span class="p">[</span><span class="s1">&#39;listeners&#39;</span><span class="p">][</span><span class="s1">&#39;mass&#39;</span><span class="p">][</span><span class="s1">&#39;cell_mass&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">2000.0</span>

<span class="c1"># run simulation and retrieve final data</span>
<span class="n">cr_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">cr_initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">chromosome_replication</span><span class="o">.</span><span class="n">topology</span><span class="p">,</span>
    <span class="s1">&#39;emit_step&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;return_raw_data&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>

<span class="n">cr_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">chromosome_replication</span><span class="p">,</span> <span class="n">cr_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: afd58b16-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:17
Completed in 0.349993 seconds
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cr_data</span><span class="p">[</span><span class="mf">0.0</span><span class="p">][</span><span class="s1">&#39;unique&#39;</span><span class="p">][</span><span class="s1">&#39;oriC&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;1&#39;: {&#39;domain_index&#39;: 1}, &#39;2&#39;: {&#39;domain_index&#39;: 2}}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cr_data</span><span class="p">[</span><span class="mf">10.0</span><span class="p">][</span><span class="s1">&#39;unique&#39;</span><span class="p">][</span><span class="s1">&#39;oriC&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;afed02fa-2025-11ec-8f9d-8c85908ac627&#39;: {&#39;domain_index&#39;: 19},
 &#39;afed0390-2025-11ec-8f9d-8c85908ac627&#39;: {&#39;domain_index&#39;: 20}}
</pre></div></div>
</div>
<p>Here a new origin of replication (oriC) has formed between time 0 and 10. This indicates the beginning of the chromosome replication process.</p>
</section>
<section id="Chromosome-Structure">
<h2>Chromosome Structure<a class="headerlink" href="#Chromosome-Structure" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.chromosome_structure.html">Chromosome Structure Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.chromosome_structure</span> <span class="kn">import</span> <span class="n">ChromosomeStructure</span>

<span class="c1"># load in parameters</span>
<span class="n">cs_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_chromosome_structure_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">chromosome_structure</span> <span class="o">=</span> <span class="n">ChromosomeStructure</span><span class="p">(</span><span class="n">cs_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">cs_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-chromosome-structure&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">chromosome</span><span class="se">\n</span><span class="s1">structure&#39;</span><span class="p">,</span>
        <span class="s1">&#39;full_chromosomes&#39;</span><span class="p">:</span> <span class="s1">&#39;full</span><span class="se">\n</span><span class="s1">chromosomes&#39;</span><span class="p">,</span>
        <span class="s1">&#39;active_ribosome&#39;</span><span class="p">:</span> <span class="s1">&#39;active</span><span class="se">\n</span><span class="s1">ribosome&#39;</span><span class="p">,</span>
        <span class="s1">&#39;chromosome_domains&#39;</span><span class="p">:</span> <span class="s1">&#39;chromosome</span><span class="se">\n</span><span class="s1">domains&#39;</span><span class="p">,</span>
        <span class="s1">&#39;active_replisomes&#39;</span><span class="p">:</span> <span class="s1">&#39;active</span><span class="se">\n</span><span class="s1">replisomes&#39;</span><span class="p">,</span>
        <span class="s1">&#39;active_RNAPs&#39;</span><span class="p">:</span> <span class="s1">&#39;active</span><span class="se">\n</span><span class="s1">RNAPs&#39;</span><span class="p">,</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">17000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">3.5</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">20</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-chromosome-structure&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">2</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">cs_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">chromosome_structure</span><span class="p">,</span> <span class="n">cs_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_51_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_51_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">cs_ports</span> <span class="o">=</span> <span class="n">chromosome_structure</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">cs_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">cs_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cs_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
listeners:
RnapData:
          n_total_collisions:
                _default: 0
                _updater: set
                _emit: True

          n_headon_collisions:
                _default: 0
                _updater: set
                _emit: True

          n_codirectional_collisions:
                _default: 0
                _updater: set
                _emit: True

          headon_collision_coordinates:
                _default: 0
                _updater: set
                _emit: True

          codirectional_collision_coordinates:
                _default: 0
                _updater: set
                _emit: True


          ... skipping 1 schema entries ...


fragmentBases:
     polymerized_ATP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     polymerized_CTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     polymerized_GTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     polymerized_UTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


molecules:
     PPI[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     WATER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     APORNAP-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


active_tfs:
     CPLX-125[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX-172[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-226[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-228[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-7669[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 19 schema entries ...

subunits:
     CPLX0-3953[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-3962[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


amino_acids:
     L-ALPHA-ALANINE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ARG[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ASN[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     L-ASPARTATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CYS[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 16 schema entries ...

active_replisomes:
*:
          domain_index:
                _default: 0
                _updater: set
                _emit: True

          coordinates:
                _default: 0
                _updater: set
                _emit: True

          unique_index:
                _default: 0
                _updater: set
                _emit: True



oriCs:
*:
          domain_index:
                _default: 0
                _updater: set
                _emit: True



chromosome_domains:
*:
          domain_index:
                _default: 0
                _updater: set
                _emit: True

          child_domains:
                _default: 0
                _updater: set
                _emit: True



active_RNAPs:
_divider:
     &lt;function divide_active_RNAPs_by_domain at 0x12dcc88b0&gt;

*:
          unique_index:
                _default: 0
                _updater: set

          domain_index:
                _default: 0
                _updater: set

          coordinates:
                _default: 0
                _updater: set

          direction:
                _default: 0
                _updater: set



RNAs:
*:
          unique_index:
                _default: 0
                _updater: set

          TU_index:
                _default: 0
                _updater: set

          transcript_length:
                _default: 0
                _updater: set

          RNAP_index:
                _default: 0
                _updater: set



active_ribosome:
*:
          protein_index:
                _default: 0

          peptide_length:
                _default: 0

          mRNA_index:
                _default: 0



full_chromosomes:
*:
          domain_index:
                _default: 0
                _updater: set
                _emit: True



promoters:
*:
          TU_index:
                _default: 0

          coordinates:
                _default: 0

          domain_index:
                _default: 0

          bound_TF:
                _default: 0



DnaA_boxes:
*:
          domain_index:
                _default: 0

          coordinates:
                _default: 0

          DnaA_bound:
                _default: 0




</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cs_initial_state</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">initial_state</span><span class="p">)</span>

<span class="c1"># run simulation and retrieve final data</span>
<span class="n">cs_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">cs_initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">chromosome_structure</span><span class="o">.</span><span class="n">topology</span><span class="p">,</span>
    <span class="s1">&#39;emit_step&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;return_raw_data&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>

<span class="n">cs_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">chromosome_structure</span><span class="p">,</span> <span class="n">cs_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: b0a8560e-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:20
Completed in 2.53 seconds
</pre></div></div>
</div>
</section>
<section id="Polypeptide-Initiation">
<h2>Polypeptide Initiation<a class="headerlink" href="#Polypeptide-Initiation" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.polypeptide_initiation.html">Polypeptide Initiation Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.polypeptide_initiation</span> <span class="kn">import</span> <span class="n">PolypeptideInitiation</span>

<span class="c1"># load in parameters</span>
<span class="n">pi_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_polypeptide_initiation_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">polypeptide_initiation</span> <span class="o">=</span> <span class="n">PolypeptideInitiation</span><span class="p">(</span><span class="n">pi_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">pi_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-polypeptide-initiation&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">polypeptide</span><span class="se">\n</span><span class="s1">initiation&#39;</span><span class="p">,</span>
        <span class="s1">&#39;active_ribosome&#39;</span><span class="p">:</span> <span class="s1">&#39;active</span><span class="se">\n</span><span class="s1">ribosome&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">17000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">3.5</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">17</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-polypeptide-initiation&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">pi_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">polypeptide_initiation</span><span class="p">,</span> <span class="n">pi_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_56_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_56_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">pi_ports</span> <span class="o">=</span> <span class="n">polypeptide_initiation</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">pi_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">pi_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pi_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
environment:
     media_id:
           _default:
           _updater: set


listeners:
ribosome_data:
          ribosomes_initialized:
                _default: 0
                _updater: set
                _emit: True

          prob_translation_per_transcript:
                _default: []
                _updater: set
                _emit: True

          effective_elongation_rate:
                _default: 0.0
                _updater: set
                _emit: True



active_ribosome:
*:
          unique_index:
                _default: 0

          protein_index:
                _default: 0

          peptide_length:
                _default: 0
                _emit: True

          mRNA_index:
                _default: 0

          pos_on_mRNA:
                _default: 0
                _emit: True



RNA:
*:
          TU_index:
                _default: 0

          can_translate:
                _default: False

          unique_index:
                _default: 0



subunits:
     CPLX0-3953[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-3962[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}



</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run simulation and retrieve final data</span>
<span class="n">pi_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">polypeptide_initiation</span><span class="o">.</span><span class="n">topology</span><span class="p">}</span>

<span class="n">pi_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">polypeptide_initiation</span><span class="p">,</span> <span class="n">pi_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: b33343ac-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:24
Completed in 2.11 seconds
</pre></div></div>
</div>
<p>We can observe the 30S ribosomal subunit (‘CPLX0-3953[c]’) and 50S ribosomal subunit (CPLX0-3962[c]) molecule counts:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot output</span>
<span class="n">pi_fig</span> <span class="o">=</span> <span class="n">plot_variables</span><span class="p">(</span>
    <span class="n">pi_data</span><span class="p">,</span>
    <span class="n">variables</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;CPLX0-3953[c]&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;CPLX0-3962[c]&#39;</span><span class="p">)</span>
        <span class="p">],</span>
    <span class="n">column_width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">row_height</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">row_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

<span class="n">pi_fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;notebooks/pi_fig.png&#39;</span><span class="p">)</span>
<span class="n">pi_fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/var/folders/t4/q28s90qs5r755xvnzjg6vkc40000gn/T/ipykernel_14901/3946241941.py:11: UserWarning: Matplotlib is currently using module://matplotlib_inline.backend_inline, which is a non-GUI backend, so cannot show the figure.
  pi_fig.show()
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_60_1.png" src="../_images/notebooks_Vivarium_ecoli_processes_60_1.png" />
</div>
</div>
<p>The decrease within the first time step of the simulation demonstrates how active 70S ribosomes are rapidly formed from free 30S and 50S subunits. We can also see how the 30S ribosome subunit (CPLX0-3953[c]) is limiting.</p>
</section>
<section id="Polypeptide-Elongation">
<h2>Polypeptide Elongation<a class="headerlink" href="#Polypeptide-Elongation" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.polypeptide_elongation.html">Polypeptide Elongation Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.polypeptide_elongation</span> <span class="kn">import</span> <span class="n">PolypeptideElongation</span>

<span class="c1"># load in parameters</span>
<span class="n">pe_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_polypeptide_elongation_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">polypeptide_elongation</span> <span class="o">=</span> <span class="n">PolypeptideElongation</span><span class="p">(</span><span class="n">pe_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">pe_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-polypeptide-elongation&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">polypeptide</span><span class="se">\n</span><span class="s1">elongation&#39;</span><span class="p">,</span>
        <span class="s1">&#39;uncharged_trna&#39;</span><span class="p">:</span> <span class="s1">&#39;uncharged_</span><span class="se">\n</span><span class="s1">trna&#39;</span><span class="p">,</span>
        <span class="s1">&#39;uncharged_trna_total&#39;</span><span class="p">:</span> <span class="s1">&#39;uncharged_</span><span class="se">\n</span><span class="s1">trna_total&#39;</span><span class="p">,</span>
        <span class="s1">&#39;charged_trna_total&#39;</span><span class="p">:</span> <span class="s1">&#39;charged_</span><span class="se">\n</span><span class="s1">trna_total&#39;</span><span class="p">,</span>
        <span class="s1">&#39;charging_molecules&#39;</span><span class="p">:</span> <span class="s1">&#39;charging_</span><span class="se">\n</span><span class="s1">molecules&#39;</span><span class="p">,</span>
        <span class="s1">&#39;active_ribosome&#39;</span><span class="p">:</span> <span class="s1">&#39;active_</span><span class="se">\n</span><span class="s1">ribosome&#39;</span><span class="p">,</span>
        <span class="s1">&#39;polypeptide_elongation&#39;</span><span class="p">:</span> <span class="s1">&#39;polypeptide</span><span class="se">\n</span><span class="s1">elongation&#39;</span><span class="p">,</span>
        <span class="s1">&#39;ppgpp_reaction_metabolites&#39;</span><span class="p">:</span> <span class="s1">&#39;ppgpp</span><span class="se">\n</span><span class="s1">reaction</span><span class="se">\n</span><span class="s1">metabolites&#39;</span><span class="p">,</span>
        <span class="s1">&#39;chromosome_domains&#39;</span><span class="p">:</span> <span class="s1">&#39;chromosome</span><span class="se">\n</span><span class="s1">domains&#39;</span><span class="p">,</span>
        <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">replication_data&#39;</span><span class="p">:</span> <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">replication_</span><span class="se">\n</span><span class="s1">data&#39;</span><span class="p">,</span>
        <span class="s1">&#39;amino_acids_total&#39;</span><span class="p">:</span> <span class="s1">&#39;amino_acids_</span><span class="se">\n</span><span class="s1">total&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">17000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">3.3</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">17</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-polypeptide-elongation&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mf">1.75</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">pe_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">polypeptide_elongation</span><span class="p">,</span> <span class="n">pe_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_64_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_64_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">pe_ports</span> <span class="o">=</span> <span class="n">polypeptide_elongation</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">pe_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">pe_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pe_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
environment:
     media_id:
           _default:
           _updater: set

amino_acids:
          L-ALPHA-ALANINE:
                _default: 0
                _divider: binomial_ecoli
                _emit: True
                _properties: {&#39;bulk&#39;: True}

          ARG:
                _default: 0
                _divider: binomial_ecoli
                _emit: True
                _properties: {&#39;bulk&#39;: True}

          ASN:
                _default: 0
                _divider: binomial_ecoli
                _emit: True
                _properties: {&#39;bulk&#39;: True}

          L-ASPARTATE:
                _default: 0
                _divider: binomial_ecoli
                _emit: True
                _properties: {&#39;bulk&#39;: True}

          CYS:
                _default: 0
                _divider: binomial_ecoli
                _emit: True
                _properties: {&#39;bulk&#39;: True}


          ... skipping 16 schema entries ...


listeners:
mass:
          cell_mass:
                _default: 0.0

          dry_mass:
                _default: 0.0


growth_limits:
          fraction_trna_charged:
                _default: 0
                _updater: set
                _emit: True

          aa_pool_size:
                _default: 0
                _updater: set
                _emit: True

          aa_request_size:
                _default: 0
                _updater: set
                _emit: True

          active_ribosomes_allocated:
                _default: 0
                _updater: set
                _emit: True

          net_charged:
                _default: []
                _updater: set
                _emit: True


          ... skipping 5 schema entries ...

ribosome_data:
          translation_supply:
                _default: 0
                _updater: set
                _emit: True

          effective_elongation_rate:
                _default: 0
                _updater: set
                _emit: True

          aaCountInSequence:
                _default: 0
                _updater: set
                _emit: True

          aaCounts:
                _default: 0
                _updater: set
                _emit: True

          actualElongations:
                _default: 0
                _updater: set
                _emit: True


          ... skipping 6 schema entries ...


molecules:
     PROTON[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     WATER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     RELA-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     SPOT-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     GUANOSINE-5DP-3DP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


molecules_total:
     PROTON[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     WATER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     RELA-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     SPOT-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     GUANOSINE-5DP-3DP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True


monomers:
     1-ACYLGLYCEROL-3-P-ACYLTRANSFER-MONOMER[i]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     1-PFK-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-DEHYDROPANTOATE-REDUCT-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-ISOPROPYLMALATESYN-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-OCTAPRENYL-METHOXY-BENZOQ-METH-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 4291 schema entries ...

amino_acids:
     L-ALPHA-ALANINE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ARG[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ASN[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     L-ASPARTATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CYS[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 16 schema entries ...

aa_enzymes:
     2-ISOPROPYLMALATESYN-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ACETOLACTSYNI-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ANTHRANSYN-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ASNSYNA-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ASPAMINOTRANS-DIMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 21 schema entries ...

ppgpp_reaction_metabolites:
     ATP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     GDP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     AMP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     GUANOSINE-5DP-3DP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     PROTON[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 2 schema entries ...

uncharged_trna:
     alaT-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     alaU-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     alaV-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     alaW-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     alaX-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 81 schema entries ...

charged_trna:
     charged-alaT-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     charged-alaU-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     charged-alaV-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     charged-alaW-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     charged-alaX-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 81 schema entries ...

charging_molecules:
     L-ALPHA-ALANINE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ATP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     alaT-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     PPI[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     AMP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 190 schema entries ...

synthetases:
     ALAS-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ARGS-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ASNS-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ASPS-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CYSS-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 16 schema entries ...

amino_acids_total:
     L-ALPHA-ALANINE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     ARG[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     ASN[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     L-ASPARTATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     CYS[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True


     ... skipping 16 schema entries ...

uncharged_trna_total:
     alaT-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     alaU-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     alaV-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     alaW-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     alaX-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True


     ... skipping 81 schema entries ...

charged_trna_total:
     charged-alaT-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     charged-alaU-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     charged-alaV-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     charged-alaW-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     charged-alaX-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True


     ... skipping 81 schema entries ...

active_ribosome:
*:
          unique_index:
                _default: 0
                _updater: set

          protein_index:
                _default: 0
                _updater: set

          peptide_length:
                _default: 0
                _updater: set
                _emit: True

          pos_on_mRNA:
                _default: 0
                _updater: set
                _emit: True

          submass:
                _default: [0. 0. 0. 0. 0. 0. 0. 0. 0.]
                _emit: True



subunits:
     CPLX0-3953[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-3962[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


polypeptide_elongation:
     aa_count_diff:
           _default: {}
           _updater: set
           _emit: True

     gtp_to_hydrolyze:
           _default: 0
           _updater: set
           _emit: True



</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run simulation and retrieve final data</span>
<span class="n">pe_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">polypeptide_elongation</span><span class="o">.</span><span class="n">topology</span><span class="p">}</span>

<span class="n">pe_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">polypeptide_elongation</span><span class="p">,</span> <span class="n">pe_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: b61fa8a8-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:29
Completed in 3.24 seconds
</pre></div></div>
</div>
<p>We can see from the cell above that each active ribosome molecule is represented by an ID number. Let’s analyze the polypeptide length and the ribosome’s position on mRNA of one active ribosome within this process:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot output</span>
<span class="n">RIBOSOME_ID</span> <span class="o">=</span> <span class="s1">&#39;1246839&#39;</span>

<span class="n">pe_fig</span> <span class="o">=</span> <span class="n">plot_variables</span><span class="p">(</span>
    <span class="n">pe_data</span><span class="p">,</span>
    <span class="n">variables</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s1">&#39;unique&#39;</span><span class="p">,</span> <span class="s1">&#39;active_ribosome&#39;</span><span class="p">,</span> <span class="n">RIBOSOME_ID</span><span class="p">,</span> <span class="s1">&#39;pos_on_mRNA&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;unique&#39;</span><span class="p">,</span> <span class="s1">&#39;active_ribosome&#39;</span><span class="p">,</span> <span class="n">RIBOSOME_ID</span><span class="p">,</span> <span class="s1">&#39;peptide_length&#39;</span><span class="p">)</span>
        <span class="p">],</span>
    <span class="n">column_width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">row_height</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">row_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_68_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_68_0.png" />
</div>
</div>
<p>Here we can see that as the simulation progresses, the ribosome travels along the mRNA strand (as shown by the increasing pos_on_mRNA variable) and polymerization of amino acids into a polypeptide occurs (as shown by the increasing peptide_length variable).</p>
<p>After elongation terminates, we can see an increase in protein counts:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pe_fig1</span> <span class="o">=</span> <span class="n">plot_variables</span><span class="p">(</span>
    <span class="n">pe_data</span><span class="p">,</span>
    <span class="n">variables</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;6PGLUCONDEHYDROG-MONOMER[c]&#39;</span><span class="p">)</span>
    <span class="p">],</span>
    <span class="n">column_width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">row_height</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">row_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_71_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_71_0.png" />
</div>
</div>
</section>
<section id="Protein-Degradation">
<h2>Protein Degradation<a class="headerlink" href="#Protein-Degradation" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.protein_degradation.html">Polypeptide Degradation Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.protein_degradation</span> <span class="kn">import</span> <span class="n">ProteinDegradation</span>

<span class="c1"># load in parameters</span>
<span class="n">pd_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_protein_degradation_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">protein_degradation</span> <span class="o">=</span> <span class="n">ProteinDegradation</span><span class="p">(</span><span class="n">pd_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">pd_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;buffer&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-protein-degradation&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">protein</span><span class="se">\n</span><span class="s1">degradation&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">10000</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-protein-degradation&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">1.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">pd_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">protein_degradation</span><span class="p">,</span> <span class="n">pd_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_74_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_74_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">pd_ports</span> <span class="o">=</span> <span class="n">protein_degradation</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">pd_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">pd_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">pd_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
metabolites:
     L-ALPHA-ALANINE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ARG[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ASN[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     L-ASPARTATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CYS[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 17 schema entries ...

proteins:
     1-ACYLGLYCEROL-3-P-ACYLTRANSFER-MONOMER[i]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     1-PFK-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-DEHYDROPANTOATE-REDUCT-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-ISOPROPYLMALATESYN-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-OCTAPRENYL-METHOXY-BENZOQ-METH-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 4291 schema entries ...


</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run simulation and retrieve final data</span>
<span class="n">pd_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">600</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">protein_degradation</span><span class="o">.</span><span class="n">topology</span><span class="p">,</span>
    <span class="s1">&#39;emit_step&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">}</span>

<span class="n">pd_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">protein_degradation</span><span class="p">,</span> <span class="n">pd_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: ba09488e-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:34
Completed in 5.69 seconds
</pre></div></div>
</div>
<p>For protein degradation, let’s look at the ARTJ-MONOMER[p] monomer, a protein selected for degradation at this time point, as it degrades into its subcomponents:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot output</span>
<span class="n">pd_fig</span> <span class="o">=</span> <span class="n">plot_variables</span><span class="p">(</span>
    <span class="n">pd_data</span><span class="p">,</span>
    <span class="n">variables</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;ARTJ-MONOMER[p]&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;ASN[c]&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;WATER[c]&#39;</span><span class="p">)</span>
        <span class="p">],</span>
    <span class="n">column_width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">row_height</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">row_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_78_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_78_0.png" />
</div>
</div>
<p>We can see here that as the ARTJ-MONOMER[p] protein degrades, the count for the amino acid asparagine increases and water is consumed.</p>
</section>
<section id="RNA-Degradation">
<h2>RNA Degradation<a class="headerlink" href="#RNA-Degradation" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.rna_degradation.html">RNA Degradation Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.rna_degradation</span> <span class="kn">import</span> <span class="n">RnaDegradation</span>

<span class="c1"># load in parameters</span>
<span class="n">rd_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_rna_degradation_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">rna_degradation</span> <span class="o">=</span> <span class="n">RnaDegradation</span><span class="p">(</span><span class="n">rd_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">rd_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-rna-degradation&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">rna</span><span class="se">\n</span><span class="s1">degradation&#39;</span><span class="p">,</span>
        <span class="s1">&#39;fragmentMetabolites&#39;</span><span class="p">:</span> <span class="s1">&#39;fragment</span><span class="se">\n</span><span class="s1">Metabolites&#39;</span><span class="p">,</span>
        <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">rna_degradation_listener&#39;</span><span class="p">:</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">listeners</span><span class="se">\n</span><span class="s1">rna_</span><span class="se">\n</span><span class="s1">degradation_</span><span class="se">\n</span><span class="s1">listener&#39;</span><span class="p">,</span>
        <span class="s1">&#39;active_ribosome&#39;</span><span class="p">:</span> <span class="s1">&#39;active_</span><span class="se">\n</span><span class="s1">ribosome&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">17000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">3.3</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">17</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-rna-degradation&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mf">1.75</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">rd_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">rna_degradation</span><span class="p">,</span> <span class="n">rd_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_82_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_82_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">rd_ports</span> <span class="o">=</span> <span class="n">rna_degradation</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">rd_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">rd_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rd_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
charged_trna:
     charged-alaT-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     charged-alaU-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     charged-alaV-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     charged-alaW-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     charged-alaX-tRNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 81 schema entries ...

bulk_RNAs:
     EG10001_RNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}
           _updater: nonnegative_accumulate

     EG10002_RNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}
           _updater: nonnegative_accumulate

     EG10003_RNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}
           _updater: nonnegative_accumulate

     EG10004_RNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}
           _updater: nonnegative_accumulate

     EG10006_RNA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}
           _updater: nonnegative_accumulate


     ... skipping 4682 schema entries ...

nmps:
     AMP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CMP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     GMP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     UMP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


fragmentMetabolites:
     polymerized_ATP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     polymerized_CTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     polymerized_GTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     polymerized_UTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     WATER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 2 schema entries ...

fragmentBases:
     polymerized_ATP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     polymerized_CTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     polymerized_GTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     polymerized_UTP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


endoRnases:
     EG10856-MONOMER[p]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10857-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10859-MONOMER[i]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10860-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10861-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 4 schema entries ...

exoRnases:
     EG11620-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     G7175-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10858-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG10863-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     EG11259-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 4 schema entries ...

subunits:
     CPLX0-3953[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX0-3962[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


molecules:
     WATER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     PROTON[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


active_ribosome:
*:
          unique_index:
                _default: 0



RNAs:
*:
          unique_index:
                _default: 0
                _updater: set

          TU_index:
                _default: 0
                _updater: set

          is_full_transcript:
                _default: 0
                _updater: set

          can_translate:
                _default: 0
                _updater: set



listeners:
mass:
          cell_mass:
                _default: 0.0

          dry_mass:
                _default: 0.0


rna_degradation_listener:
          fraction_active_endo_rnases:
                _default: 0
                _updater: set
                _emit: True

          diff_relative_first_order_decay:
                _default: 0
                _updater: set
                _emit: True

          fract_endo_rrna_counts:
                _default: 0
                _updater: set
                _emit: True

          count_rna_degraded:
                _default: 0
                _updater: set
                _emit: True

          nucleotides_from_degradation:
                _default: 0
                _updater: set
                _emit: True


          ... skipping 1 schema entries ...



</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run simulation and retrieve final data</span>
<span class="n">rd_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">rna_degradation</span><span class="o">.</span><span class="n">topology</span><span class="p">}</span>

<span class="n">rd_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">rna_degradation</span><span class="p">,</span> <span class="n">rd_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: bdf352a0-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:42
Completed in 10.27 seconds
</pre></div></div>
</div>
<p>For RNA degradation, let’s look at a few RNA molecules as the simulation runs:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot output</span>
<span class="n">rd_fig</span> <span class="o">=</span> <span class="n">plot_variables</span><span class="p">(</span>
    <span class="n">rd_data</span><span class="p">,</span>
    <span class="n">variables</span><span class="o">=</span><span class="p">[</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;RNA0-300[c]&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;RNA0-301[c]&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;bulk&#39;</span><span class="p">,</span> <span class="s1">&#39;WATER[c]&#39;</span><span class="p">)</span>
        <span class="p">],</span>
    <span class="n">column_width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">row_height</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">row_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_86_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_86_0.png" />
</div>
</div>
<p>RNAs are selected and degraded by endoRNases, and non-functional RNA fragments are digested through exoRNases. During the process water is consumed, and nucleotides, pyrophosphate and protons are released.</p>
</section>
<section id="Two-Component-System">
<h2>Two Component System<a class="headerlink" href="#Two-Component-System" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.two_component_system.html">Two Component System Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.two_component_system</span> <span class="kn">import</span> <span class="n">TwoComponentSystem</span>

<span class="c1"># load in parameters</span>
<span class="n">tcs_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_two_component_system_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">two_component_system</span> <span class="o">=</span> <span class="n">TwoComponentSystem</span><span class="p">(</span><span class="n">tcs_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">tcs_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-two-component-system&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli two</span><span class="se">\n</span><span class="s1">component</span><span class="se">\n</span><span class="s1">system&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">16000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">5.0</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">18</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-two-component-system&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">1.35</span><span class="p">,</span> <span class="mi">1</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">tcs_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">two_component_system</span><span class="p">,</span> <span class="n">tcs_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_90_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_90_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">tcs_ports</span> <span class="o">=</span> <span class="n">two_component_system</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">tcs_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">tcs_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tcs_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
molecules:
     PHOSPHO-ARCB-CPLX[i]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ADP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     PROTON[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ARCB-CPLX[i]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ATP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 36 schema entries ...

listeners:
mass:
          cell_mass:
                _default: 0




</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># tweak initial state??</span>
<span class="n">tcs_initial_state</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">initial_state</span><span class="p">)</span>

<span class="c1"># run simulation and retrieve final data</span>
<span class="n">tcs_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">two_component_system</span><span class="o">.</span><span class="n">topology</span><span class="p">}</span>

<span class="n">tcs_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">two_component_system</span><span class="p">,</span> <span class="n">tcs_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: c62fb3f0-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:55
Completed in 0.038426 seconds
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcs_data</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;bulk&#39;: {&#39;PHOSPHO-ARCB-CPLX[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;ADP[c]&#39;: [404873, 404873, 404873, 404873, 404873, 404873],
  &#39;PROTON[c]&#39;: [52, 52, 52, 52, 52, 52],
  &#39;ARCB-CPLX[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;ATP[c]&#39;: [7866030, 7866030, 7866030, 7866030, 7866030, 7866030],
  &#39;PHOSPHO-ARCA[c]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;ARCA-MONOMER[c]&#39;: [2769, 2769, 2769, 2769, 2769, 2769],
  &#39;Pi[c]&#39;: [408412, 408412, 408412, 408412, 408412, 408412],
  &#39;WATER[c]&#39;: [34912886892,
   34912886892,
   34912886892,
   34912886892,
   34912886892,
   34912886892],
  &#39;PHOSPHO-BAES-INDOLE-CPLX[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;BAES-INDOLE-CPLX[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;PHOSPHO-BAER[c]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;BAER-MONOMER[c]&#39;: [200, 200, 200, 200, 200, 200],
  &#39;PHOSPHO-BAES[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;BAES-MONOMER[i]&#39;: [19, 19, 19, 19, 19, 19],
  &#39;PHOSPHO-BASS-FE+3-CPLX[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;BASS-FE+3-CPLX[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;PHOSPHO-BASR[c]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;BASR-MONOMER[c]&#39;: [180, 180, 180, 180, 180, 180],
  &#39;PHOSPHO-BASS[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;BASS-MONOMER[i]&#39;: [46, 46, 46, 46, 46, 46],
  &#39;PHOSPHO-DCUS-SUC-CPLX[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;DCUS-SUC-CPLX[p]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;PHOSPHO-DCUR[c]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;DCUR-MONOMER[c]&#39;: [80, 80, 80, 80, 80, 80],
  &#39;PHOSPHO-DCUS[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;DCUS-MONOMER[p]&#39;: [12, 12, 12, 12, 12, 12],
  &#39;PHOSPHO-NARX-NITRATE-CPLX[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;NARX-NITRATE-CPLX[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;PHOSPHO-NARL[c]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;NARL-MONOMER[c]&#39;: [679, 679, 679, 679, 679, 679],
  &#39;PHOSPHO-NARX[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;NARX-CPLX[i]&#39;: [17, 17, 17, 17, 17, 17],
  &#39;PHOSPHO-PHOQ[i]&#39;: [1, 1, 1, 1, 1, 1],
  &#39;CPLX0-8168[i]&#39;: [13, 13, 13, 13, 13, 13],
  &#39;PHOSPHO-PHOP[c]&#39;: [148, 148, 148, 148, 148, 148],
  &#39;PHOP-MONOMER[c]&#39;: [855, 855, 855, 855, 855, 855],
  &#39;PHOSPHO-PHOR[i]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;PHOR-CPLX[c]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;PHOSPHO-PHOB[c]&#39;: [0, 0, 0, 0, 0, 0],
  &#39;PHOB-MONOMER[c]&#39;: [88, 88, 88, 88, 88, 88]},
 &#39;listeners&#39;: {&#39;mass&#39;: {}},
 &#39;time&#39;: [0.0, 2.0, 4.0, 6.0, 8.0, 10.0]}
</pre></div></div>
</div>
<p>Phosphate groups are transferred from histidine kinases to response regulators and back in response to counts of ligand stimulants</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[62]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># # plot output</span>
<span class="c1"># tcs_fig = plot_variables(</span>
<span class="c1">#     tcs_data,</span>
<span class="c1">#     variables=[</span>
<span class="c1">#         ],</span>
<span class="c1">#     column_width=10, row_height=3, row_padding=0.5)</span>
</pre></div>
</div>
</div>
</section>
<section id="Equilibrium">
<h2>Equilibrium<a class="headerlink" href="#Equilibrium" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.equilibrium.html">Equilibrium Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[63]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.equilibrium</span> <span class="kn">import</span> <span class="n">Equilibrium</span>

<span class="c1"># load in parameters</span>
<span class="n">eq_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_equilibrium_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">equilibrium</span> <span class="o">=</span> <span class="n">Equilibrium</span><span class="p">(</span><span class="n">eq_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[64]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">eq_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-equilibrium&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">equilibrium&#39;</span><span class="p">,</span>
        <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">equilibrium_listener&#39;</span><span class="p">:</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">listeners</span><span class="se">\n</span><span class="s1">equilibrium</span><span class="se">\n</span><span class="s1">listener&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">14000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">5.0</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">18</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-equilibrium&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">1.5</span><span class="p">,</span> <span class="mf">1.25</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">eq_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">equilibrium</span><span class="p">,</span> <span class="n">eq_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_98_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_98_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[65]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">eq_ports</span> <span class="o">=</span> <span class="n">equilibrium</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">eq_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">eq_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">eq_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
molecules:
     CPLX-123[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     PC00033[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     HYPOXANTHINE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CPLX-125[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     TRP[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 93 schema entries ...

listeners:
mass:
          cell_mass:
                _default: 0


equilibrium_listener:
          reaction_rates:
                _default: []
                _updater: set
                _emit: True




</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[66]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run simulation and retrieve final data</span>
<span class="n">eq_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">equilibrium</span><span class="o">.</span><span class="n">topology</span><span class="p">}</span>

<span class="n">eq_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">equilibrium</span><span class="p">,</span> <span class="n">eq_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: c64fab9c-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:55
Completed in 0.182374 seconds
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># # plot output</span>
<span class="c1"># eq_fig = plot_variables(</span>
<span class="c1">#     eq_data,</span>
<span class="c1">#     variables=[</span>
<span class="c1">#         ],</span>
<span class="c1">#     column_width=10, row_height=3, row_padding=0.5)</span>
</pre></div>
</div>
</div>
</section>
<section id="Metabolism">
<h2>Metabolism<a class="headerlink" href="#Metabolism" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="https://covertlab.github.io/vivarium-ecoli/reference/api/ecoli.processes.metabolism.html">Metabolism Documentation</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[68]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ecoli.processes.metabolism</span> <span class="kn">import</span> <span class="n">Metabolism</span>

<span class="c1"># load in parameters</span>
<span class="n">meta_params</span> <span class="o">=</span> <span class="n">load_sim_data</span><span class="o">.</span><span class="n">get_metabolism_config</span><span class="p">()</span>

<span class="c1"># initialize process and topology</span>
<span class="n">metabolism</span> <span class="o">=</span> <span class="n">Metabolism</span><span class="p">(</span><span class="n">meta_params</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[69]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot topology</span>
<span class="n">meta_topology_plot_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;node_labels&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;ecoli-metabolism&#39;</span><span class="p">:</span> <span class="s1">&#39;ecoli</span><span class="se">\n</span><span class="s1">metabolism&#39;</span><span class="p">,</span>
        <span class="s1">&#39;kinetics_enzymes&#39;</span><span class="p">:</span> <span class="s1">&#39;kinetics</span><span class="se">\n</span><span class="s1">enzymes&#39;</span><span class="p">,</span>
        <span class="s1">&#39;kinetics_substrates&#39;</span><span class="p">:</span> <span class="s1">&#39;kinetics</span><span class="se">\n</span><span class="s1">substrates&#39;</span><span class="p">,</span>
        <span class="s1">&#39;environment</span><span class="se">\n</span><span class="s1">exchange_data&#39;</span><span class="p">:</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">environment</span><span class="se">\n</span><span class="s1">exchange_</span><span class="se">\n</span><span class="s1">data&#39;</span><span class="p">,</span>
        <span class="s1">&#39;listeners</span><span class="se">\n</span><span class="s1">enzyme_kinetics&#39;</span><span class="p">:</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">listeners</span><span class="se">\n</span><span class="s1">enzyme_</span><span class="se">\n</span><span class="s1">kinetics&#39;</span><span class="p">,</span>
        <span class="s1">&#39;polypeptide_elongation&#39;</span><span class="p">:</span> <span class="s1">&#39;polypeptide_</span><span class="se">\n</span><span class="s1">elongation&#39;</span>
    <span class="p">},</span>
    <span class="s1">&#39;show_ports&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
    <span class="s1">&#39;node_size&#39;</span><span class="p">:</span> <span class="mi">15000</span><span class="p">,</span>
    <span class="s1">&#39;node_distance&#39;</span><span class="p">:</span> <span class="mf">3.2</span><span class="p">,</span>
    <span class="s1">&#39;dashed_edges&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s1">&#39;font_size&#39;</span><span class="p">:</span> <span class="mi">17</span><span class="p">,</span>
    <span class="s1">&#39;graph_format&#39;</span><span class="p">:</span> <span class="s1">&#39;hierarchy&#39;</span><span class="p">,</span>
    <span class="s1">&#39;coordinates&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;ecoli-metabolism&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">4.5</span><span class="p">,</span> <span class="mi">2</span><span class="p">)}</span>
<span class="p">}</span>

<span class="n">meta_topology_fig</span> <span class="o">=</span> <span class="n">plot_topology</span><span class="p">(</span><span class="n">metabolism</span><span class="p">,</span> <span class="n">meta_topology_plot_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_104_0.png" src="../_images/notebooks_Vivarium_ecoli_processes_104_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[70]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># display ports schema</span>
<span class="n">meta_ports</span> <span class="o">=</span> <span class="n">metabolism</span><span class="o">.</span><span class="n">ports_schema</span><span class="p">()</span>
<span class="n">meta_printout</span> <span class="o">=</span> <span class="n">make_port_printout</span><span class="p">(</span><span class="n">meta_ports</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">meta_printout</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
metabolites:
     2-3-DIHYDROXYBENZOATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-KETOGLUTARATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-PG[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2K-4CH3-PENTANOATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     4-AMINO-BUTYRATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 167 schema entries ...

catalysts:
     1-ACYLGLYCEROL-3-P-ACYLTRANSFER-MONOMER[i]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     1-PFK[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-DEHYDROPANTOATE-REDUCT-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-ISOPROPYLMALATESYN-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-OCTAPRENYL-METHOXY-BENZOQ-METH-MONOMER[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 1472 schema entries ...

kinetics_enzymes:
     3-ISOPROPYLMALDEHYDROG-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     6PFK-1-CPX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     6PFK-2-CPX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     7-ALPHA-HYDROXYSTEROID-DEH-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     7KAPSYN-CPLX[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 307 schema entries ...

kinetics_substrates:
     2-3-DIHYDROXYBENZOATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     2-KETOGLUTARATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     4-AMINO-BUTYRATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     4-hydroxybenzoate[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ACETYL-COA[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 86 schema entries ...

amino_acids:
     L-ALPHA-ALANINE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ARG[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     ASN[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     L-ASPARTATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}

     CYS[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True
           _properties: {&#39;bulk&#39;: True}


     ... skipping 16 schema entries ...

amino_acids_total:
     L-ALPHA-ALANINE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     ARG[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     ASN[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     L-ASPARTATE[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True

     CYS[c]:
           _default: 0
           _divider: binomial_ecoli
           _emit: True


     ... skipping 16 schema entries ...

environment:
     media_id:
           _default:
           _updater: set

exchange:
          ACET[p]:
                _default: 0

          AMMONIUM[c]:
                _default: 0

          BETAINE[p]:
                _default: 0

          BUTANAL[c]:
                _default: 0

          CA+2[p]:
                _default: 0


          ... skipping 46 schema entries ...

exchange_data:
          unconstrained:
                _default: []

          constrained:
                _default: []



listeners:
mass:
          cell_mass:
                _default: 0.0

          dry_mass:
                _default: 0.0


fba_results:
          media_id:
                _default:
                _updater: set

          conc_updates:
                _default: []
                _updater: set

          catalyst_counts:
                _default: 0
                _updater: set

          translation_gtp:
                _default: 0
                _updater: set

          coefficient:
                _default: 0.0
                _updater: set


          ... skipping 12 schema entries ...

enzyme_kinetics:
          metaboliteCountsInit:
                _default: 0
                _updater: set

          metaboliteCountsFinal:
                _default: 0
                _updater: set

          enzymeCountsInit:
                _default: 0
                _updater: set

          countsToMolar:
                _default: 1.0
                _updater: set

          actualFluxes:
                _default: []
                _updater: set


          ... skipping 3 schema entries ...


polypeptide_elongation:
     aa_count_diff:
           _default: {}
           _emit: True

     gtp_to_hydrolyze:
           _default: 0
           _emit: True



</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[71]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run simulation and retrieve final data</span>
<span class="n">meta_settings</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;total_time&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
    <span class="s1">&#39;initial_state&#39;</span><span class="p">:</span> <span class="n">initial_state</span><span class="p">,</span>
    <span class="s1">&#39;topology&#39;</span><span class="p">:</span> <span class="n">metabolism</span><span class="o">.</span><span class="n">topology</span><span class="p">}</span>

<span class="n">meta_data</span> <span class="o">=</span> <span class="n">simulate_process</span><span class="p">(</span><span class="n">metabolism</span><span class="p">,</span> <span class="n">meta_settings</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

Simulation ID: c73b86d4-2025-11ec-8f9d-8c85908ac627
Created: 09/27/2021 at 23:31:57
Completed in 1.80 seconds
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[72]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reaction_fluxes</span> <span class="o">=</span> <span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;listeners&#39;</span><span class="p">][</span><span class="s1">&#39;fba_results&#39;</span><span class="p">][</span><span class="s1">&#39;reactionFluxes&#39;</span><span class="p">]</span>
<span class="k">del</span> <span class="n">reaction_fluxes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># generate time series</span>
<span class="n">reaction_fluxes_time_series</span> <span class="o">=</span> <span class="p">{}</span>

<span class="k">for</span> <span class="n">fluxes</span> <span class="ow">in</span> <span class="n">reaction_fluxes</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">reaction_id</span><span class="p">,</span> <span class="n">flux</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fluxes</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">reaction_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">reaction_fluxes_time_series</span><span class="p">:</span>
            <span class="n">reaction_fluxes_time_series</span><span class="p">[</span><span class="n">reaction_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">reaction_fluxes_time_series</span><span class="p">[</span><span class="n">reaction_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">flux</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[74]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># plot figure</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>

<span class="k">for</span> <span class="n">reaction_id</span><span class="p">,</span> <span class="n">time_series</span> <span class="ow">in</span> <span class="n">reaction_fluxes_time_series</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">time_series</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Reaction Fluxes&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[74]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Text(0.5, 1.0, &#39;Reaction Fluxes&#39;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Vivarium_ecoli_processes_109_1.png" src="../_images/notebooks_Vivarium_ecoli_processes_109_1.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../tutorials.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="Vivarium_ecoli_composites.html" class="btn btn-neutral float-right" title="Vivarium E. coli Composites" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2023, The Vivarium E. coli Authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>